File: app/src/main/java/com/awscherb/cardkeeper/ui/cards/CardsFragment.java
package com.awscherb.cardkeeper.ui.cards;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import com.awscherb.cardkeeper.R;
import com.awscherb.cardkeeper.dagger.component.DaggerCardsComponent;
import com.awscherb.cardkeeper.dagger.module.CardsPresenterModule;
import com.awscherb.cardkeeper.data.model.ScannedCode;
import com.awscherb.cardkeeper.ui.base.BaseApplication;
import com.awscherb.cardkeeper.ui.base.BaseFragment;
import com.awscherb.cardkeeper.ui.scan.ScanActivity;
import com.awscherb.cardkeeper.ui.scan.ScanFragment;
import com.google.zxing.BarcodeFormat;

import java.util.List;

import javax.inject.Inject;

import static android.app.Activity.RESULT_OK;

public class CardsFragment extends BaseFragment implements CardsContract.View {

    public static final int REQUEST_GET_CODE = 3;

    @Inject
    CardsContract.Presenter presenter;
    RecyclerView recyclerView;
    LinearLayoutManager layoutManager;
    CardsAdapter scannedCodeAdapter;
    FloatingActionButton fab;

    //================================================================================
    // New Instance
    //================================================================================

    public static CardsFragment newInstance() {
        return new CardsFragment();
    }

    //================================================================================
    // Lifecycle methods
    //================================================================================

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        DaggerCardsComponent.builder()
                .cardsPresenterModule(new CardsPresenterModule(this))
                .servicesComponent(((BaseApplication) getActivity().getApplication()).getServicesComponent())
                .build().inject(this);

    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_cards, container, false);
        if (v == null) {
            return null;
        }

        recyclerView = (RecyclerView) v.findViewById(R.id.fragment_cards_recycler);
        fab = (FloatingActionButton) v.findViewById(R.id.fragment_cards_fab);

        layoutManager = new LinearLayoutManager(getActivity());
        scannedCodeAdapter = new CardsAdapter(getActivity(), presenter);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(scannedCodeAdapter);

        fab.setOnClickListener(v1 -> startActivityForResult(
                new Intent(getActivity(), ScanActivity.class), REQUEST_GET_CODE));

        return v;
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.loadCards();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == REQUEST_GET_CODE && resultCode == RESULT_OK) {
            ScannedCode scannedCode = new ScannedCode();
            scannedCode.setId(System.currentTimeMillis());
            scannedCode.setText(data.getStringExtra(ScanFragment.EXTRA_BARCODE_TEXT));
            scannedCode.setFormat((BarcodeFormat) data.getSerializableExtra(ScanFragment.EXTRA_BARCODE_FORMAT));

            EditText input = new EditText(getActivity());
            input.setHint(R.string.dialog_card_name_hint);
            input.setInputType(InputType.TYPE_TEXT_FLAG_CAP_WORDS);

            new AlertDialog.Builder(getActivity())
                    .setTitle(R.string.app_name)
                    .setView(input)
                    .setPositiveButton(R.string.action_add,
                            (dialog, which) -> {
                                scannedCode.setTitle(input.getText().toString());
                                presenter.addNewCard(scannedCode);
                            })
                    .setNegativeButton(R.string.action_cancel,
                            (dialog, which) -> dialog.dismiss())
                    .show();
        }
    }

    //================================================================================
    // View methods
    //================================================================================

    @Override
    public void showCards(List<ScannedCode> codes) {
        scannedCodeAdapter.swapObjects(codes);
    }

}

File: app/src/main/java/com/awscherb/cardkeeper/ui/cards/CardsFragment.kt
package com.awscherb.cardkeeper.ui.cards

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.activityViewModels
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.awscherb.cardkeeper.R
import com.awscherb.cardkeeper.ui.base.BaseFragment
import com.awscherb.cardkeeper.ui.base.CardKeeperNavigator
import com.awscherb.cardkeeper.ui.card_detail.CardDetailViewModel
import com.awscherb.cardkeeper.ui.card_detail.CardDetailViewModelFactory
import com.google.android.material.floatingactionbutton.FloatingActionButton
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import javax.inject.Inject

class CardsFragment : BaseFragment() {

    private val viewModel by activityViewModels<CardsViewModel> { viewModelFactory }

    @Inject
    lateinit var viewModelFactory: CardsViewModelFactory

    private val detailViewModel by activityViewModels<CardDetailViewModel> { detailFactory }

    @Inject
    lateinit var detailFactory: CardDetailViewModelFactory

    private val navigator by lazy { CardKeeperNavigator(resources.getBoolean(R.bool.portrait)) }

    private lateinit var toolbar: Toolbar
    private lateinit var searchLayout: View
    private lateinit var searchBox: EditText
    private lateinit var searchClose: ImageView
    private lateinit var recyclerView: RecyclerView
    private lateinit var fab: FloatingActionButton

    private lateinit var layoutManager: LinearLayoutManager
    private lateinit var scannedCodeAdapter: CardsAdapter

    //================================================================================
    // Lifecycle methods
    //================================================================================

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = inflater.inflate(R.layout.fragment_cards, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewComponent.inject(this)

        toolbar = view.findViewById(R.id.fragment_cards_toolbar)
        recyclerView = view.findViewById(R.id.fragment_cards_recylcer)
        searchLayout = view.findViewById(R.id.fragment_cards_search_layout)
        searchBox = view.findViewById(R.id.fragment_cards_search)
        searchClose = view.findViewById(R.id.fragment_cards_search_close)

        fab = view.findViewById(R.id.fragment_cards_fab)

        setupRecycler()
        setupToolbar()

        fab.setOnClickListener {
            findNavController().navigate(
                CardsFragmentDirections.actionCardsFragmentToScanFragment()
            )
        }

        viewModel.cards.onEach {
            scannedCodeAdapter.items = it
        }.launchIn(lifecycleScope)

        searchClose.setOnClickListener {
            dismissKeyboard()
            searchBox.setText("")
            searchLayout.visibility = View.GONE
            toolbar.visibility = View.VISIBLE
        }

        searchBox.addLifecycleTextWatcher {
            viewModel.searchQuery.value = it
        }
    }

    //================================================================================
    // View methods
    //================================================================================

    private fun onCardDeleted() {
        showSnackbar(R.string.fragment_cards_deleted_card)
    }

    //================================================================================
    // Helper methods
    //================================================================================

    private fun setupToolbar() {
        toolbar.inflateMenu(R.menu.menu_cards)
        toolbar.setOnMenuItemClickListener {
            when (it.itemId) {
                R.id.create -> {
                    findNavController().navigate(R.id.action_cardsFragment_to_createFragment)
                    true
                }
                R.id.search -> {
                    toolbar.visibility = View.GONE
                    searchBox.requestFocusAndShowKeyboard()
                    searchLayout.visibility = View.VISIBLE
                    true
                }
                else -> false
            }
        }
    }

    private fun setupRecycler() {
        layoutManager = GridLayoutManager(activity, resources.getInteger(R.integer.cards_columns))
        scannedCodeAdapter = CardsAdapter(requireActivity(), {
            detailViewModel.cardId.value = it.id
            navigator.navigateToDetail(this, it.id)
        }) { code ->
            AlertDialog.Builder(requireContext())
                .setTitle(R.string.adapter_scanned_code_delete_message)
                .setPositiveButton(R.string.action_delete) { _, _ ->
                    viewModel.deleteCard(code)
                    onCardDeleted()
                }
                .setNegativeButton(R.string.action_cancel, null)
                .show()
        }
        recyclerView.layoutManager = layoutManager
        recyclerView.adapter = scannedCodeAdapter
    }
}
